event-sourcing/event-store-postgres/src/main/java/org/apache/james/eventsourcing/eventstore/postgres/PostgresEventStoreDAO.java

```diff
@@ -0,0 +1,122 @@
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.EVENT_ID;
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.SNAPSHOT;
+import static org.apache.james.eventsourcing.eventstore.postgres.PostgresEventStoreModule.PostgresEventStoreTable.TABLE_NAME;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.eventsourcing.AggregateId;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.EventId;
+import org.apache.james.eventsourcing.eventstore.History;
+import org.apache.james.util.ReactorUtils;
+import org.jooq.JSON;
+import org.jooq.Record;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.jdk.javaapi.CollectionConverters;
+
+public class PostgresEventStoreDAO {
+    private PostgresExecutor postgresExecutor;
+    private JsonEventSerializer jsonEventSerializer;
+
+    @Inject
+    public PostgresEventStoreDAO(PostgresExecutor postgresExecutor, JsonEventSerializer jsonEventSerializer) {
+        this.postgresExecutor = postgresExecutor;
+        this.jsonEventSerializer = jsonEventSerializer;
+    }
+
+    public Mono<Void> appendAll(List<Event> events, Optional<EventId> lastSnapshot) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, AGGREGATE_ID, EVENT_ID, EVENT)
+                .valuesOfRecords(events.stream().map(event -> dslContext.newRecord(AGGREGATE_ID, EVENT_ID, EVENT)
+                    .value1(event.getAggregateId().asAggregateKey())
+                    .value2(event.eventId().serialize())
+                    .value3(convertToJooqJson(event))).collect(ImmutableList.toImmutableList()))))
+            .then(lastSnapshot.map(eventId -> insertSnapshot(events.iterator().next().getAggregateId(), eventId)).orElse(Mono.empty()));
+    }
+
+    private Mono<Void> insertSnapshot(AggregateId aggregateId, EventId snapshotId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+            .set(SNAPSHOT, snapshotId.serialize())
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))));
+    }
+
+    private JSON convertToJooqJson(Event event) {
+        try {
+            return JSON.json(jsonEventSerializer.serialize(event));
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public Mono<EventId> getSnapshot(AggregateId aggregateId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.select(SNAPSHOT)
+            .from(TABLE_NAME)
+            .where(AGGREGATE_ID.eq(aggregateId.asAggregateKey()))
```

Comment:
And SNAPSHOT IS NOT NULL

+ an index I bet

Reply:
There is no "SNAPSHOT IS NOT NULL" in Cassandra implement:

```
session.prepare(QueryBuilder
      .selectFrom(EVENTS_TABLE)
      .column(SNAPSHOT)
      .whereColumn(AGGREGATE_ID).isEqualTo(bindMarker(AGGREGATE_ID))
      .build())
```

Reply:
index added